### PR TITLE
AWS: Support source_profile

### PR DESCRIPTION
**Pull request by AndyTitu**
## Overview
This change makes so that having `source_profile` in your `.aws/config` file doesn't error out and instead behaves as expected.

**This is the expected behaviour:**

1. Case 1: Only `source_profile` is specified in a profile. --> The source profile is inspected recursively and the right credentials provider is chosen to be used for the current profile. 

2. Case 2: `source_profile` is specified, alongside other role/mfa specific information (`role_arn`. `mfa_serial` etc) --> the source profile is inspected recursively to determine the right provider. This provider is then used as "base" credentials for executing SDK calls to the STS service for retrieving another set of temporary credentials as marked by the other role/mfa specific information present in the current profile.

**Further aspects regarding recursivity:**

- Base case for recursivity is using the access keys stored in the associated 1Password item.
- Endless loops caused by `source_profile` are handled when parsing the config file using was-vault's `LoadFromProfile` function.

## Type of change
 
- [x] Improved an existing plugin

## Related Issue(s)
  

* Resolves: #212 

## How To Test
Use a config file that specifies `source_profile` and execute `aws sts get-caller-identity --profile <name>` to check that the right credentials were retrieved.

Example of config file:
```
[default]
source_profile=andy
output=json
region=us-east-1

[profile andy]
role_arn=arn:aws:iam::123456789012:role/testRole
mfa_serial=arn:aws:iam::123456789012:mfa/andi

[profile dev]
source_profile=Andy
role_arn=arn:aws:iam::123456789012:role/testRole1

[profile prod]
source_profile=dev
role_arn=arn:aws:iam::123456789012:role/testRole2
```

## Changelog
AWS Shell Plugin now supports sourcing credentials from another profile.

### CODE REVIEW

**Comment by AndyTitu** [comment]
> What's the source of the expected behaviour as described in the PR description? How do we know that is actually how this works?

According to https://docs.aws.amazon.com/cli/latest/topic/config-vars.html#using-aws-iam-roles:

> source_profile - The AWS CLI profile that contains credentials / configuration the CLI should use for the initial assume-role call. This profile may be another profile configured to use assume-role, though if static credentials are present in the profile they will take precedence.

**Comment by AndyTitu** [comment]
> Could we add a test to make sure that a source_profile loop is detected? That would make sure that we still catch that even if we replace the code to load the config file.

On a second check, I have realised there's a bug in `aws-vault` logic which makes it not detect endless loops in `source_profile`, whereas the AWS cli detects them. In order to not rewrite all logic for parsing config files, as we are using `aws-vault`'s util function, I have added an additional check which is performed before continuing to parse the config file. 

I've also found out self-loops, i.e. profiles which are referencing themselves, are accepted as they would normally work with AWS cli, and it's our tools responsibility to gracefully handle that case. `aws-vault` util on which we rely, handles that.

I have introduced a couple of regression tests to cover both roundtrip loop and self-loops.

**Comment by AndyTitu** [comment]
There still is another edge case matter I'm looking to address: Using mfa authentication in multiple of the chained profiles. Specifically using the same virtual MFA device in multiple profiles in the chain. 

Using several different MFA devices along the chain of profiles is out of scope because we only support 1 item per plugin.

File: plugins/aws/sts_provisioner.go [hunk 1]
```diff
@@ -231,11 +243,7 @@ func resolveLocalAnd1PasswordConfigurations(itemFields map[sdk.FieldName]string,
 	}
 
 	if awsConfig.HasMfaSerial() && awsConfig.MfaToken == "" {
-		return fmt.Errorf("MFA failed: an MFA serial was found but no OTP has been set up in 1Password")
-	}
-
-	if !awsConfig.HasMfaSerial() && awsConfig.MfaToken != "" {
-		return fmt.Errorf("MFA failed: an OTP was found wihtout a corresponding MFA serial")
+		return fmt.Errorf("MFA failed: the selected profile requires an OTP because an MFA serial (%s) was detected but no 'One-Time Password' field was found in the associated item", awsConfig.MfaSerial)
```
Comment (AndyTitu) [inline plugins/aws/sts_provisioner.go:246]: @Lucy (couldn't find your official handle) do error messages like this sound fine? The question applies for all error messages in this function, which users could be seeing while setting things up.

Comment (libutcher) [reply plugins/aws/sts_provisioner.go:246]: Am I understanding correctly that this error would be returned when an MFA serial is found on the 1P item, but no one-time password field is found?

If so, I wonder if it would be more helpful to explain the actions to take to fix this (adding an OTP or removing the serial) than to explain that the existence of the MFA serial prompted this.

I'm not sure if we typically include docs links in error messages, so here are two options with or without a learn more link. Please feel free to edit these as needed! 

MFA failed: an MFA serial (%s) was detected on the associated item for the selected profile, but no 'One-Time Password' field was found. Learn how to add an OTP field to your item: https://developer.1password.com/docs/cli/shell-plugins/aws/#optional-set-up-multi-factor-authentication Or remove the MFA serial field to proceed without multi-factor authentication.

MFA failed: an MFA serial (%s) was detected on the associated item for the selected profile, but no 'One-Time Password' field was found. Add an OTP to your item to use multi-factor authentication, or remove the MFA serial field to proceed without MFA.

File: plugins/aws/access_key_test.go [hunk 2]
```diff
@@ -392,6 +417,132 @@ func TestSTSProvisioner(t *testing.T) {
...
+		"WithEndlessLoop": {
+			ItemFields: map[sdk.FieldName]string{
+				fieldname.AccessKeyID:     "AKIAHPIZFMD5EEXAMPLE",
+				fieldname.SecretAccessKey: "lBfKB7P5ScmpxDeRoFLZvhJbqNGPoV0vIEXAMPLE",
+				fieldname.DefaultRegion:   "us-central-1",
+				fieldname.OneTimePassword: "908789",
+				fieldname.MFASerial:       "arn:aws:iam::123456789012:mfa/user1",
+			},
+			ExpectedOutput: sdk.ProvisionOutput{
+				Diagnostics: sdk.Diagnostics{Errors: []sdk.Error{{Message: "infinite loop in credential configuration detected. Attempting to load from profile \"prod\" which has already been visited"}}},
```
Comment (libutcher) [inline plugins/aws/access_key_test.go:517]: @AndyTitu I'm not sure I understand what "has already been visited" in this error message means. Does it mean it's defaulting to another profile that has been used before?

File: plugins/aws/sts_provisioner.go [hunk 3]
```diff
@@ -355,3 +360,30 @@ func getSTSClient(region string, credsProvider aws.CredentialsProvider) *sts.Cli
...
+	for sourceProfile != "" {
+		if visited[sourceProfile] {
+			return fmt.Errorf("infinite loop in credential configuration detected. Attempting to load from profile %q which has already been visited", sourceProfile)
+		} else {
+			visited[sourceProfile] = true
+		}
+
+		profileSection, ok := config.ProfileSection(sourceProfile)
+		if !ok {
+			return fmt.Errorf("source profile %s does not exist in the config file", sourceProfile)
```
Comment (libutcher) [inline plugins/aws/sts_provisioner.go:377]: Should we specify AWS config file?

```suggestion
			return fmt.Errorf("source profile %s does not exist in your AWS config file.", sourceProfile)
```

**Review by libutcher — COMMENTED** [review]
Took a look at the rest of the error messages and left some small questions and notes, but overall I think they look good!

File: plugins/aws/access_key_test.go (hunk 2 above)
Comment (AndyTitu) [reply plugins/aws/access_key_test.go:517]: Yes, while assuming an aws profile, the user could specify that they'd want that profile to source credentials from another profile. To get the right credentials, the plugin needs to visit source profiles as declared in the config file. If a profile was visited twice it means that an endless cycle was detected.

Comment (AndyTitu) [reply plugins/aws/access_key_test.go:517]: I have copy pasted this error message from the AWS CLI which outputs the exact same text for this issue

File: plugins/aws/sts_provisioner.go (hunk 3 above)
Comment (AndyTitu) [reply plugins/aws/sts_provisioner.go]: Done thanks!

File: plugins/aws/sts_provisioner.go (hunk 1 above)
Comment (AndyTitu) [reply plugins/aws/sts_provisioner.go]: Done, thanks!

File: plugins/aws/access_key_test.go (hunk 2 above)
Comment (libutcher) [reply plugins/aws/access_key_test.go:517]: Ah okay, I'm sure it makes sense to regular AWS CLI users in that case!

File: plugins/aws/sts_provisioner.go [hunk 4]
```diff
@@ -355,3 +360,30 @@ func getSTSClient(region string, credsProvider aws.CredentialsProvider) *sts.Cli
...
+	for sourceProfile != "" {
+		if visited[sourceProfile] {
+			return fmt.Errorf("infinite loop in credential configuration detected. Attempting to load from profile %q which has already been visited", sourceProfile)
+		} else {
+			visited[sourceProfile] = true
+		}
+
+		profileSection, ok := config.ProfileSection(sourceProfile)
+		if !ok {
+			return fmt.Errorf("source profile %s does not exist in your AWS config file", sourceProfile)
```
Comment (jpcoenen) [inline plugins/aws/sts_provisioner.go:377]: The profile is user input, so let's use `%q` here.
```suggestion
			return fmt.Errorf("source profile %q does not exist in your AWS config file", sourceProfile)
```

**Review by jpcoenen — APPROVED** [review]
One tiny remark, but overall this looks good to me. Great job, Andy! 🙌

File: plugins/aws/sts_provisioner.go [hunk 5]
```diff
@@ -231,11 +243,7 @@ func resolveLocalAnd1PasswordConfigurations(itemFields map[sdk.FieldName]string,
 	}
 
 	if awsConfig.HasMfaSerial() && awsConfig.MfaToken == "" {
-		return fmt.Errorf("MFA failed: an MFA serial was found but no OTP has been set up in 1Password")
-	}
-
-	if !awsConfig.HasMfaSerial() && awsConfig.MfaToken != "" {
-		return fmt.Errorf("MFA failed: an OTP was found wihtout a corresponding MFA serial")
+		return fmt.Errorf("MFA failed: an MFA serial (%s) was detected on the associated item for the selected profile, but no 'One-Time Password' field was found. Add an OTP to your item to use multi-factor authentication", awsConfig.MfaSerial)
```
Comment (hculea) [inline plugins/aws/sts_provisioner.go:246]: Let's rephrase this to something more readable:
```suggestion
		return fmt.Errorf("MFA failed: MFA serial %q was detected on the associated item for the selected profile, but no 'One-Time Password' field was found. Add an OTP to your item to use multi-factor authentication", awsConfig.MfaSerial)
```

Comment (hculea) [reply plugins/aws/sts_provisioner.go:246]: Also, this may be a good place to link a docs link to how to add an OTP to the item, it's not something self explanatory.

File: plugins/aws/sts_provisioner.go [hunk 6]
```diff
@@ -97,9 +92,14 @@ func (p STSProvisioner) Description() string {
 }
 
 // ChooseTemporaryCredentialsProvider returns the aws provider that fits the scenario described by the current configuration.
-func ChooseTemporaryCredentialsProvider(awsConfig *confighelpers.Config, providerFactory STSProviderFactory) (aws.CredentialsProvider, error) {
+func ChooseTemporaryCredentialsProvider(awsConfig *confighelpers.Config, providerFactory STSProviderFactory, itemFields map[sdk.FieldName]string) (aws.CredentialsProvider, error) {
```
Comment (hculea) [inline plugins/aws/sts_provisioner.go:95]: nit: you can consider omitting "Choose" from this function's name, as it does not add any value.

File: plugins/aws/sts_provisioner.go [hunk 7]
```diff
@@ -97,9 +92,14 @@ func (p STSProvisioner) Description() string {
...
+	err := resolveLocalAnd1PasswordConfigurations(itemFields, awsConfig)
+	if err != nil {
+		return nil, err
+	}
```
Comment (hculea) [inline plugins/aws/sts_provisioner.go:99]: what is the reason this check is now required, but it was not before?

**Review by hculea — COMMENTED** [review]
Did a code-review, looks good in broad lines, only a couple nits and a question!

File: plugins/aws/sts_provisioner.go (hunk 5 above)
Comment (AndyTitu) [reply plugins/aws/sts_provisioner.go:246]: Do we generally add links to error messages? It would make this specific error message rather long:

```
MFA failed: an MFA serial (%s) was detected on the associated item for the selected profile, but no 'One-Time Password' field was found. Learn how to add an OTP field to your item: https://developer.1password.com/docs/cli/shell-plugins/aws/#optional-set-up-multi-factor-authentication

```
(as per Lucy's suggestion)

File: plugins/aws/sts_provisioner.go (hunk 7 above)
Comment (AndyTitu) [reply plugins/aws/sts_provisioner.go:99]: This was required before as well, I just moved it over from another function inside here, so it would be performed for each profile, because now profiles can look into each other recursively, by calling function `ChooseTemporaryCredentialsProvider ` recursively. Before it was called in the caller function of `ChooseTemporaryCredentialsProvider` so it would resolve fields only for first profile in the chain.

File: plugins/aws/sts_provisioner.go (hunk 5 above)
Comment (hculea) [reply plugins/aws/sts_provisioner.go:246]: I'd definitely add some spacing in there, for better readability:
```
MFA failed: an MFA serial (%s) was detected on the associated item for the selected profile, but no 'One-Time Password' field was found. 

Learn how to add an OTP field to your item: 

https://developer.1password.com/docs/cli/shell-plugins/aws/#optional-set-up-multi-factor-authentication
```

Comment (hculea) [reply plugins/aws/sts_provisioner.go:246]: > Do we generally add links to error messages?

not consistently, although I believe we should do this. Not providing an actionable way for the user to proceed, after getting an error message, is very frustrating and can be a bottleneck for retention

File: plugins/aws/sts_provisioner.go (hunk 6 above)
Comment (AndyTitu) [reply plugins/aws/sts_provisioner.go]: Done

File: plugins/aws/sts_provisioner.go (hunk 4 above)
Comment (AndyTitu) [reply plugins/aws/sts_provisioner.go]: Done, thanks!

File: plugins/aws/sts_provisioner.go (hunk 5 above)
Comment (AndyTitu) [reply plugins/aws/sts_provisioner.go]: Okay, added it. Done!

File: plugins/aws/access_key_test.go [hunk 8]
```diff
@@ -469,17 +620,7 @@ func TestResolveLocalAnd1PasswordConfigurations(t *testing.T) {
...
-		{
-			description: "has mfa token but no mfa serial",
-			itemFields: map[sdk.FieldName]string{
-				fieldname.OneTimePassword: "515467",
-			},
-			awsConfig: &confighelpers.Config{
-				ProfileName: "dev",
-			},
-			err: fmt.Errorf("MFA failed: an OTP was found wihtout a corresponding MFA serial"),
+			err: fmt.Errorf("MFA failed: MFA serial \"arn:aws:iam::123456789012:mfa/user\" was detected on the associated item or in the config file for the selected profile, but no 'One-Time Password' field was found. Learn how to add an OTP field to your item: https://developer.1password.com/docs/cli/shell-plugins/aws/#optional-set-up-multi-factor-authentication"),
```
Comment (hculea) [inline plugins/aws/access_key_test.go:623]: Any issues around adding some new spacing here?
```suggestion
			err: fmt.Errorf(`MFA failed: MFA serial \"arn:aws:iam::123456789012:mfa/user\" was detected on the associated item or in the config file for the selected profile, but no 'One-Time Password' field was found.

Learn how to add an OTP field to your item:

https://developer.1password.com/docs/cli/shell-plugins/aws/#optional-set-up-multi-factor-authentication`),

```

**Review by hculea — APPROVED** [review]

File: plugins/aws/access_key_test.go (hunk 8 above)
Comment (AndyTitu) [reply plugins/aws/access_key_test.go]: Done, thanks!

**Review by hculea — APPROVED** [review]
